metal/mutexkv.go

```diff
@@ -0,0 +1,51 @@
+package metal
+
+import (
+	"log"
+	"sync"
+)
+
+// MutexKV is a simple key/value store for arbitrary mutexes. It can be used to
+// serialize changes across arbitrary collaborators that share knowledge of the
+// keys they must serialize on.
+//
+// The initial use case is to let aws_security_group_rule resources serialize
```

Comment:
Perhaps we could change this to note why we are using it in this provider:
```
               // Equinix Metal doesn't allow multiple VLANs to be assigned
                // to the same port at the same time
                lockId := "vlan-attachment-" + port.ID
                metalMutexKV.Lock(lockId)
                defer metalMutexKV.Unlock(lockId)
```

I wonder if we really need this. A mutex lock could be used in other places, I think, but perhaps a backoff-retry loop  would be sufficient and save us the complexity of locks.

Is this even more important in newer TF because of provider aliases?

Reply:
looks like the latest text failed related to these changes:
`metal/provider.go:9:20: undefined: mutexkv`

Reply:
> A mutex lock could be used in other places, I think, but perhaps a backoff-retry loop would be sufficient and save us the complexity of locks.

I really think that a mutex lock and unlock is much more simple than a retry loop. It's just 3 lines, no error checking, no API calls.

> Is locking even more important in newer TF because of provider aliases?

I don't understand how provider aliases are related. The locking here is to prevent TF parallelism to ruin an apply run - TF runs several goroutines and parallelizes everyhting that is not dependent. I think a HC eng did this after acceptance test were failing because of parallel port-vlan assignment. I am not sure if the mutex map is the best solution, but we didn't have an issue with the parallel assignment since.

I had to drop the mutex map code because it was removed in the SDK v2.
https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#removal-of-helper-mutexkv-package

Reply:
I assumed the mutex would be shared between all aliases of the provider, but now I'm not sure about that. With aliases perhaps each instance is executing a separate copy of the plugin binary.

A retry loop would prevent cases where multiple clients (not just terraform, or multiple copies of terraform) are running against the same account. Maybe something to consider in the future, maybe not.

Agreed - Updating the mutex for sdkv2 sounds good.